File: csharp/Almondbread/Complex.cs
using System;

namespace Almondbread
{
    internal struct Complex
    {
        internal static Complex ZERO = new Complex();
        private double real;
        private double imag;
        private double abs;

        internal Complex(double real = 0, double imag = 0)
        {
            this.real = real;
            this.imag = imag;
            this.abs = Math.Sqrt(real * real + imag * imag);
        }

        public static Complex operator +(Complex a, Complex b)
        {
            return new Complex(a.real + b.real, a.imag + b.imag);
        }

        public static Complex operator +(Complex a, double b)
        {
            return new Complex(a.real + b, a.imag);
        }

        public static Complex operator +(double b, Complex a)
        {
            return new Complex(a.real + b, a.imag);
        }

        public static Complex operator *(Complex a, Complex b)
        {
            return new Complex(a.real * b.real - a.imag * b.imag,
                    a.imag * b.real + a.real * b.imag);
        }

        public static implicit operator double(Complex c)
        {
            return c.abs;
        }
    }
}

File: java/src/main/java/almondbread/Complex.java
package almondbread;

final class Complex {
    final static Complex ZERO = new Complex();
    final double real;
    final double imag;
    final double abs;

    Complex(final double real, final double imag) {
        this.real = real;
        this.imag = imag;
        this.abs = Math.sqrt(real * real + imag * imag);
    }

    Complex(final double real) {
        this.real = real;
        this.imag = 0;
        this.abs = real;
    }

    Complex() {
        this.real = 0;
        this.imag = 0;
        this.abs = 0;
    }

    final Complex add(final Complex other) {
        return new Complex(real + other.real, imag + other.imag);
    }

    final Complex add(final double other) {
        return new Complex(real + other, imag);
    }

    final Complex mul(final Complex other) {
        return new Complex(real * other.real - imag * other.imag,
                imag * other.real + real * other.imag);
    }
}
